File: frontend/src/types/canonicalSkill.ts
import type { CanonicalRecord } from "./canonicalBase";
import type { CanonicalModifierReference } from "./canonicalModifier";

export interface CanonicalSkill extends CanonicalRecord {
  owner_class_ids: string[];
  owner_mastery_ids: string[];
  source_skill_id?: string | null;
  skill_tree_id?: string | null;
  skill_tags: string[];
  damage_types: string[];
  scaling_tags: string[];
  cost?: Record<string, unknown>;
  cast_data?: Record<string, unknown>;
  cooldown?: Record<string, unknown>;
  requirements?: Record<string, unknown>;
  behavior_flags?: Record<string, unknown>;
  description_text?: string | null;
  tooltip_text?: string | null;
  text_effects: string[];
  special_behavior_classification?: string;
  damage_source_status?: string | null;
  stable_calculable?: boolean;
}

export interface CanonicalSkillTreeNode extends CanonicalRecord {
  skill_tree_id?: string | null;
  skill_id?: string | null;
  owner_class_ids?: string[];
  owner_mastery_ids?: string[];
  source_tree_id?: string | null;
  source_node_id?: number | string | null;
  max_points?: number | null;
  required_points?: number | null;
  position: Record<string, number>;
  connections?: string[];
  edge_requirements?: Array<{ node_id: string; points: number }>;
  node_type?: string | null;
  modifier_references: CanonicalModifierReference[];
  modifier_rows?: Array<Record<string, unknown>>;
  description_text?: string | null;
  tooltip_text?: string | null;
  text_effects: string[];
  special_behavior_classification?: string;
  effect_hint_classifications?: string[];
  stable_calculable?: boolean;
}

export interface CanonicalSkillTree extends CanonicalRecord {
  skill_id?: string | null;
  owner_class_ids?: string[];
  owner_mastery_ids?: string[];
  source_tree_id?: string | null;
  source_tree_class_name?: string | null;
  node_ids: string[];
  edges: Array<[string, string]>;
  stable_calculable?: boolean;
}

File: backend/app/data_contracts/canonical_skill.py
"""Canonical skill and skill tree contracts."""

from __future__ import annotations

from dataclasses import dataclass, field

from .canonical_base import CanonicalRecord
from .canonical_modifier import CanonicalModifierReference


@dataclass(frozen=True)
class CanonicalSkill(CanonicalRecord):
    owner_class_ids: tuple[str, ...] = field(default_factory=tuple)
    owner_mastery_ids: tuple[str, ...] = field(default_factory=tuple)
    skill_tree_id: str | None = None
    tags: tuple[str, ...] = field(default_factory=tuple)
    damage_types: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "owner_class_ids", tuple(self.owner_class_ids))
        object.__setattr__(self, "owner_mastery_ids", tuple(self.owner_mastery_ids))
        object.__setattr__(self, "tags", tuple(self.tags))
        object.__setattr__(self, "damage_types", tuple(self.damage_types))


@dataclass(frozen=True)
class CanonicalSkillTreeNode(CanonicalRecord):
    skill_tree_id: str | None = None
    max_points: int | None = None
    required_points: int | None = None
    position: dict[str, float] = field(default_factory=dict)
    modifier_references: tuple[CanonicalModifierReference, ...] = field(default_factory=tuple)
    text_effects: tuple[str, ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "position", dict(self.position))
        object.__setattr__(self, "modifier_references", tuple(self.modifier_references))
        object.__setattr__(self, "text_effects", tuple(self.text_effects))


@dataclass(frozen=True)
class CanonicalSkillTree(CanonicalRecord):
    skill_id: str | None = None
    node_ids: tuple[str, ...] = field(default_factory=tuple)
    edges: tuple[tuple[str, str], ...] = field(default_factory=tuple)

    def __post_init__(self) -> None:
        super().__post_init__()
        object.__setattr__(self, "node_ids", tuple(self.node_ids))
        object.__setattr__(self, "edges", tuple(tuple(edge) for edge in self.edges))
